hpy/devel/src/runtime/argparse.c

```diff
@@ -94,75 +214,95 @@ HPyAPI_RUNTIME_FUNC(int)
 HPyArg_ParseKeywords(HPyContext ctx, HPy *args, HPy_ssize_t nargs, HPy kw,
                      const char *fmt, const char *keywords[], ...)
 {
-  va_list vl;
-  va_start(vl, keywords);
-  const char *fmt1 = fmt;
-  int optional = 0;
-  int keyword_only = 0;
-  HPy_ssize_t i = 0;
-  HPy_ssize_t nkw = 0;
-  HPy current_arg;
-
-  // first count positional only arguments
-  while (keywords[nkw] != NULL && !*keywords[nkw]) nkw++;
-  // then check and count the rest
-  while (keywords[nkw] != NULL) {
-    if (!*keywords[nkw]) {
-      HPyErr_SetString(ctx, ctx->h_TypeError, "XXX: Empty keyword parameter name");
-      va_end(vl);
-      return 0;
+    const char *fmt1 = fmt;
+    const char *err_funcname = NULL;
+    const char *err_message = NULL;
+    const char *fmt_end = NULL;
+
+    int optional = 0;
+    int keyword_only = 0;
+    HPy_ssize_t i = 0;
+    HPy_ssize_t nkw = 0;
+    HPy current_arg;
+
+    fmt_end = parse_errstring(fmt, &err_funcname, &err_message);
+
+    // first count positional only arguments
+    while (keywords[nkw] != NULL && !*keywords[nkw]) nkw++;
+    // then check and count the rest
+    while (keywords[nkw] != NULL) {
+        if (!*keywords[nkw]) {
+            set_error(ctx, ctx->h_SystemError, err_funcname, err_message,
+                "empty keyword parameter name");
+            return 0;
+        }
+        nkw++;
     }
-    nkw++;
-  }
-
-  while (*fmt1 != 0) {
-      if (*fmt1 == '|') {
-        optional = 1;
-        fmt1++;
-        continue;
-      }
-      if (*fmt1 == '$') {
-        optional = 1;
-        keyword_only = 1;
-        fmt1++;
-        continue;
-      }
-      if (i >= nkw) {
-        HPyErr_SetString(ctx, ctx->h_TypeError, "XXX: mismatched args (too few keywords for fmt)");
-        va_end(vl);
-        return 0;
-      }
-      current_arg = HPy_NULL;
-      if (i < nargs) {
-        if (keyword_only) {
-          HPyErr_SetString(ctx, ctx->h_TypeError, "XXX: keyword only argument passed as positional argument");
-          va_end(vl);
-          return 0;
-        }
-        current_arg = args[i];
-      }
-      else if (!HPy_IsNull(kw) && *keywords[i]) {
-        current_arg = HPyDict_GetItem(ctx, kw, HPyUnicode_FromString(ctx, keywords[i]));
-      }
-      if (!HPy_IsNull(current_arg) || optional) {
-        if (!_HPyArg_ParseItem(ctx, current_arg, &fmt1, &vl)) {
-          va_end(vl);
-          return 0;
-        }
-      }
-      else {
-        HPyErr_SetString(ctx, ctx->h_TypeError, "XXX: no value for required argument");
+
+    va_list vl;
+    va_start(vl, keywords);
+
+    while (fmt1 != fmt_end) {
+        if (*fmt1 == '|') {
```

Comment:
Personally, I'd extract each condition to a named helper function as it makes the code clearer and requires fewer comments to understand.
They'll be inlined anyway.

I don't feel strongly about this one though.

Reply:
I'm happy for these couple of special cases to stay, but I do want to split up the pieces of `parse_item`.